File: longestCommonSubstring.js
// Write a function, longest_common_substring(str1, str2) that takes two strings and returns the longest common substring.
// A substring is defined as any consecutive slice of letters from another string.

// Bonus: solve it in O(m * n) using O(m * n) extra space.
// (Hint: the solution involves dynamic programming which will be introduced later in the course.)

const longest_common_substring = function(str1, str2) {
  firstSubstrings = subStrings(str1);
  secondSubstrings = subStrings(str2);
  let length = 0;
  for (let i = 0; i < firstSubstrings.length; i++) {
    let sub = firstSubstrings[i];
    if (secondSubstrings.includes(sub) && sub.length > length) {
      length = sub.length;
    }
  }
  return length;
};

const subStrings = function(str) {
  let subs = [];
  let chars = str.split("");
  for (let i = 0; i < chars.length; i++) {
    for (let j = i + 1; j < chars.length; j++) {
      let sub = chars.slice(i, j + 1);
      subs.push(sub.join(""));
    }
  }
  return subs;
};

const makeMatrix = function (arr) {
    
}

// console.log(/subStrings("abcdefghijk"));

console.log(longest_common_substring("abcdefgh", "xxavcdef"));

File: longest_common_substring.rb
def make_matrix(str1, str2)
matrix = Array.new(str1.length + 1) {Array.new(str2.length + 1, 0)}
    str1.chars.each_with_index do |el1,idx1|
        str2.chars.each_with_index do |el2, idx2|
        if el1 == el2 
            matrix[idx1 + 1][idx2 + 1] = matrix[idx1][idx2] + 1
        else
            matrix[idx1 + 1][idx2 + 1] = 0;
        end
    end
end
matrix
end

def longest_common_substring(str1, str2)
matrix = make_matrix(str1,str2)
greatest_substring = ""
    matrix.each_with_index do |row, idx1|
        row.each_with_index do |length, idx2|
            if length > greatest_substring.length
                greatest_substring = str1[idx2 - length...idx2]
            end
        end
    end
    greatest_substring
end


#    |"" | C | A | T  
# --------------------
# "" | 0 | 0 | 0 | 0
# --------------------
# R  | 0 | 0 | 0 | 0
# --------------------
# A  | 0 | 0 | 1 | 0
# --------------------
# T  | 0 | 0 | 0 | 2 